### PR TITLE
Use 5 bits to encode setting words

### PR DESCRIPTION
**Pull request by fiam**
Saves 1672 bits on F3 without any lost functionality. Retriving a
setting name is significantly slower, but still not slow enough to be
problematic.

### CODE REVIEW

File: src/utils/settings.rb [hunk 1]
```diff
@@ -455,13 +467,62 @@ def write_impl_file(file)
...
+            if c == 0
+                chr = 0 # XXX: Remove this if we go for explicit lengths
+            elsif c >= 'a'.ord && c <= 'z'.ord
+                chr = 1 + (c - 'a'.ord)
+            elsif c >= 'A'.ord && c <= 'Z'.ord
+                raise "Cannot encode uppercase character #{c.ord} (#{c})"
+            else
+                idx = symbols.index(c)
+                if idx.nil?
+                    if rem_symbols == 0
```
Comment (alexbirkett) [inline src/utils/settings.rb:490]: Is `rem_symbols` ever decremented?

**Comment by digitalentity** [comment]
Are 5 bits enough? What if we add more settings?

File: src/utils/settings.rb (hunk 1 above)
Comment (fiam) [reply src/utils/settings.rb:490]: Nice catch! That’s definitely a bug, I’ll fix it.

**Comment by fiam** [comment]
This exploits the fact that most setting names use only a-z characters. This lets use use 5 bits encoding for those as long as we don’t use more then 5 characters outside that range (we need to encode the zero char too). IIRC, we’re using 3 characters outside that range (numbers 1, 2 and 3). If we need more than 31 characters we can switch it to 6 bits, which still saves space but I don’t remember exactly how much. I’ll test it later and I’ll post the memory usage with 6 bit encoding.

**Comment by fiam** [comment]
Moved the number of bits to a constant at the top of `settings.rb` and removed it from `settings.c` by making the settings compiler output it in `settings_generated.h`.

**Pull request merged**